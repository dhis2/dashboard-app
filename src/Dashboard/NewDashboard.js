import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import DashboardVerticalOffset from './DashboardVerticalOffset';
import EditBar from '../ControlBarContainer/EditBar';
import DashboardContent from './DashboardContent';
import { acSetEditNewDashboard } from '../actions/editDashboard';

class NewDashboard extends Component {
    componentDidMount() {
        this.props.setNewDashboard();
    }

    render() {
        return (
            <Fragment>
                <EditBar />
                <DashboardVerticalOffset />
                <div className="dashboard-wrapper">
                    <DashboardContent editMode={true} />
                </div>
            </Fragment>
        );
    }
}

export default connect(
    null,
    {
        setNewDashboard: acSetEditNewDashboard,
    }
)(NewDashboard);
