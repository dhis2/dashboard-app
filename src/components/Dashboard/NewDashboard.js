import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';

import { acSetEditNewDashboard } from '../../actions/editDashboard';
import DashboardVerticalOffset from './DashboardVerticalOffset';
import EditBar from '../ControlBar/EditBar';
import DashboardContent from './DashboardContent';

class NewDashboard extends Component {
    componentDidMount() {
        this.props.setNewDashboard();
    }

    render() {
        return (
            <Fragment>
                <EditBar />
                <DashboardVerticalOffset editMode={true} />
                <div className="dashboard-wrapper">
                    <DashboardContent editMode={true} />
                </div>
            </Fragment>
        );
    }
}

export default connect(
    null,
    {
        setNewDashboard: acSetEditNewDashboard,
    }
)(NewDashboard);
