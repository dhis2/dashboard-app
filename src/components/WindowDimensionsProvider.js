import React, { createContext, useContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import debounce from 'lodash/debounce'

export const WindowDimensionsCtx = createContext(null)

const windowDims = () => ({
    height: window.innerHeight,
    width: window.innerWidth,
})

const WindowDimensionsProvider = ({ children }) => {
    const [dimensions, setDimensions] = useState(windowDims())

    useEffect(() => {
        const handleResize = debounce(() => {
            setDimensions(windowDims())
        }, 200)
        window.addEventListener('resize', handleResize)
        return () => {
            window.removeEventListener('resize', handleResize)
        }
    }, [])

    return (
        <WindowDimensionsCtx.Provider value={dimensions}>
            {children}
        </WindowDimensionsCtx.Provider>
    )
}

WindowDimensionsProvider.propTypes = {
    children: PropTypes.node.isRequired,
}

export default WindowDimensionsProvider

export const useWindowDimensions = () => {
    return useContext(WindowDimensionsCtx)
}
