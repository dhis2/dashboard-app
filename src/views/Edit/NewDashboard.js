import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import cx from 'classnames'

import DashboardContainer from '../../components/DashboardContainer'
import ActionsBar from './ActionsBar'
import TitleBar from './TitleBar'
import ItemGrid from './ItemGrid'
import LayoutPrintPreview from '../print/PrintLayoutDashboard'
import NotSupportedNotice from './NotSupportedNotice'

import { acSetEditNewDashboard } from '../../actions/editDashboard'
import { sGetIsPrintPreviewView } from '../../reducers/editDashboard'

import classes from './styles/NewDashboard.module.css'

const NewDashboard = props => {
    useEffect(() => {
        props.setNewDashboard()
    }, [])

    return (
        <>
            <div
                className={cx(classes.container, 'dashboard-scroll-container')}
            >
                <ActionsBar />
                {props.isPrintPreviewView ? (
                    <LayoutPrintPreview fromEdit={true} />
                ) : (
                    <DashboardContainer>
                        <TitleBar />
                        <ItemGrid />
                    </DashboardContainer>
                )}
            </div>
            <div className={classes.notice}>
                <NotSupportedNotice
                    className={classes.notSupportedNotice}
                    message={i18n.t(
                        'Creating dashboards on small screens is not supported. Resize your screen to return to create mode.'
                    )}
                />
            </div>
        </>
    )
}

NewDashboard.propTypes = {
    isPrintPreviewView: PropTypes.bool,
    setNewDashboard: PropTypes.func,
}

const mapStateToProps = state => ({
    isPrintPreviewView: sGetIsPrintPreviewView(state),
})

export default connect(mapStateToProps, {
    setNewDashboard: acSetEditNewDashboard,
})(NewDashboard)
