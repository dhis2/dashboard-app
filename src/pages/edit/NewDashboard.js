import i18n from '@dhis2/d2-i18n'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { useState, useEffect } from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import { acSetEditNewDashboard } from '../../actions/editDashboard'
import DashboardContainer from '../../components/DashboardContainer'
import Notice from '../../components/Notice'
import { useWindowDimensions } from '../../components/WindowDimensionsProvider'
import { setHeaderbarVisible } from '../../modules/setHeaderbarVisible'
import { isSmallScreen } from '../../modules/smallScreen'
import { sGetIsPrintPreviewView } from '../../reducers/editDashboard'
import LayoutPrintPreview from '../print/PrintLayoutDashboard'
import ActionsBar from './ActionsBar'
import ItemGrid from './ItemGrid'
import classes from './styles/NewDashboard.module.css'
import TitleBar from './TitleBar'

const NewDashboard = props => {
    const { width } = useWindowDimensions()
    const [redirectUrl, setRedirectUrl] = useState(null)

    useEffect(() => {
        if (isSmallScreen(width)) {
            setRedirectUrl('/')
            return
        }
        setHeaderbarVisible(true)
        props.setNewDashboard()
    }, [])

    if (redirectUrl) {
        return <Redirect to={redirectUrl} />
    }

    return (
        <>
            <div
                className={cx(classes.container, 'dashboard-scroll-container')}
            >
                <ActionsBar />
                {props.isPrintPreviewView ? (
                    <LayoutPrintPreview fromEdit={true} />
                ) : (
                    <DashboardContainer>
                        <TitleBar />
                        <ItemGrid />
                    </DashboardContainer>
                )}
            </div>
            <div className={classes.notice}>
                <Notice
                    title={i18n.t('Not supported')}
                    message={i18n.t(
                        'Creating dashboards on small screens is not supported. Resize your screen to return to create mode.'
                    )}
                />
            </div>
        </>
    )
}

NewDashboard.propTypes = {
    isPrintPreviewView: PropTypes.bool,
    setNewDashboard: PropTypes.func,
}

const mapStateToProps = state => ({
    isPrintPreviewView: sGetIsPrintPreviewView(state),
})

export default connect(mapStateToProps, {
    setNewDashboard: acSetEditNewDashboard,
})(NewDashboard)
