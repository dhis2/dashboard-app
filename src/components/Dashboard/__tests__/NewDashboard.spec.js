import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import WindowDimensionsProvider from '../../WindowDimensionsProvider'
import NewDashboard from '../NewDashboard'

jest.mock(
    '../../TitleBar/TitleBar',
    () =>
        function MockTitleBar() {
            return <div>TitleBar</div>
        }
)
jest.mock(
    '../../ItemGrid/ItemGrid',
    () =>
        function MockItemGrid() {
            return <div>ItemGrid</div>
        }
)

jest.mock(
    '../PrintLayoutDashboard',
    () =>
        function MockLayoutPrintPreview() {
            return <div>LayoutPrintPreview</div>
        }
)

const mockStore = configureMockStore()

const store = {
    dashboards: {
        byId: {
            rainbowdash: {
                id: 'rainbowdash',
                access: {
                    update: true,
                    delete: true,
                },
            },
        },
        items: [],
    },
    selected: {
        id: 'rainbowdash',
    },
    windowHeight: 900,
    editDashboard: {
        id: '',
        access: {},
        printPreviewView: false,
    },
}

test('NewDashboard renders dashboard', () => {
    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <NewDashboard />
            </WindowDimensionsProvider>
        </Provider>
    )

    expect(container).toMatchSnapshot()
})

test('NewDashboard renders print preview', () => {
    store.editDashboard.printPreviewView = true

    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <NewDashboard />
            </WindowDimensionsProvider>
        </Provider>
    )
    expect(container).toMatchSnapshot()
})
