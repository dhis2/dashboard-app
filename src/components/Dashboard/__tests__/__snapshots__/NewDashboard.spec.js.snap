// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewDashboard renders dashboard 1`] = `
<div>
  <div
    class="editBar"
  >
    <div
      class="controls"
    >
      <div
        class="jsx-71743532 start"
        data-test="dhis2-uicore-buttonstrip"
      >
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 primary"
            data-test="save-dashboard-button"
            type="button"
          >
            Save changes
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Print preview
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        />
        <div
          class="jsx-71743532 box"
        />
      </div>
      <button
        class="jsx-2371629422 secondary"
        data-test="dhis2-uicore-button"
        type="button"
      >
        Exit without saving
      </button>
    </div>
  </div>
  <div
    class="dashboard-wrapper"
    style="height: 788px;"
  >
    <div>
      TitleBar
    </div>
    <div>
      ItemGrid
    </div>
  </div>
</div>
`;

exports[`NewDashboard renders print preview 1`] = `
<div>
  <div
    class="editBar"
  >
    <div
      class="controls"
    >
      <div
        class="jsx-71743532 start"
        data-test="dhis2-uicore-buttonstrip"
      >
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 primary"
            data-test="save-dashboard-button"
            type="button"
          >
            Save changes
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Exit Print preview
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        />
        <div
          class="jsx-71743532 box"
        />
      </div>
      <button
        class="jsx-2371629422 secondary"
        data-test="dhis2-uicore-button"
        type="button"
      >
        Exit without saving
      </button>
    </div>
  </div>
  <div>
    LayoutPrintPreview
  </div>
</div>
`;
